File: binarysearch.io/October 2020/725_counting_maximal_value_roots_in_binary_tree.cpp
#include "solution.hpp"
using namespace std;

/**
 * class Tree {
 *     public:
 *         int val;
 *         Tree *left;
 *         Tree *right;
 * };
 */
class Solution {
    public:
    int val = 0;
    int solve(Tree* root) {
        helper(root);
        return val;
    }
    int helper(Tree* root){
        if(root == NULL){
            return 0;
        }
        if(root->left == NULL && root->right == NULL){
            val += 1;
            return root->val;
        }
        int v1 = helper(root->left);
        int v2 = helper(root->right);
        int m = max(v1,v2);
        m = max(root->val,m);
        if (m == root->val){
            val += 1;
        }
        return m;
    }
};

File: binarysearch.io/725_counting_maximal_value_roots_in_binary_tree.py
# class Tree:
#     def __init__(self, val, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, root):
        def helper(root):
            if root is None:
                return 0,-float('inf')
            if root.left is None and root.right is None:
                print(1,root.val)
                return 1,root.val
            
            c1,v1 = helper(root.left)
            c2,v2 = helper(root.right)
            count = c1 + c2 
            
            m = max(v1,v2,root.val)
            if m <= root.val:
                count += 1 
            print(count,root.val)
            return count,m 
        
        return helper(root)[0] 
